feat(expansion-panel): allow for content to be rendered lazily

crisbeto opened the pull request:
Currently the expansion panel renders all of its content on load which can be expensive, considering that all of it won't be visible. These changes allow for some of the content to be rendered lazily via the `matExpansionPanelContent` directive.

Fixes #8230.

josephperrott commented on src/lib/expansion/expansion-panel.ts:
```diff
@@ -127,12 +147,31 @@ export class MatExpansionPanel extends _MatExpansionPanelMixinBase
+      }
+    });
+
+    // Remove the content once the panel is closed and the closing animation is done.
```
What is the reasoning behind removing the content on closing?  Since the issue we are looking to address is lazy loading, is there a reason to detach and reattach all the time?

crisbeto replied:
I was a bit on the fence on this one as well, but it could help if you have some really massive panels with lots of elements. I'm open to not doing this though.

bardiel replied:
Could this be configured perhaps via an Input?
I'm developing (for learning purposes) an RSS feed reader. On desktop, each item is placed inside an expansion panel (for fancyness (?)); inspired by Google Inbox interface.

Removing the content inside the panel after been read, could be something I'd like to happen. Being that as more and more items are been loaded, more resources will be consumed. (O course I shall do something about this regardless of this specific feature).

However, also thinking about other Google products like GSuit admin console: where some settings are loaded asynchronously the first time, but then kept. (I don't know if they're actually removed from DOM, but I wouldn't remove them in such scenario).

josephperrott replied:
I think it might be best at this point to not remove the content once it is placed and we can determine if it is something that we want to allow later, potentially as @bardiel mentioned via `@Input`

josephperrott commented on src/lib/expansion/expansion-panel.ts:
```diff
@@ -127,12 +147,31 @@ export class MatExpansionPanel extends _MatExpansionPanelMixinBase
     return this.expanded ? 'expanded' : 'collapsed';
   }
 
+  ngAfterViewInit() {
```
We have had a number of issues with content changed after checked for these situations, are we going to run into the same issue here?

crisbeto replied:
Probably not, those exceptions are usually if data bindings change. This is the same way we're doing it for tabs.

josephperrott replied:
For tabs we actually are having to change it because of these issues #7266 

We are creating a custom template portal and having the template portal itself be responsible for attachment because of issues with `ChangeAfterContentChecked`.  I just want to be sure we don't accidentally create an issue further down the road.

jelbourn commented on src/lib/expansion/expansion-panel.html:
```diff
@@ -1,8 +1,12 @@
+     [id]="id">
   <div class="mat-expansion-panel-body">
-    <ng-content></ng-content>
+    <ng-template><ng-content></ng-content></ng-template>
```
This doesn't stop Angular from initializing stuff; we can make this change to stop stuff from making it into the DOM, but we'll ultimately need an alternate API where users provide an `ng-template`

crisbeto replied:
Ah, I was under the impression that this was the reason we do it the way we do for the tabs. In that case I'll rework it to use a `@ContentChild` to find an `ng-template` and instantiate it.

amcdnl replied:
@crisbeto - Look at what I did for https://github.com/angular/material2/pull/6832

crisbeto commented:
Rebased and reworked the approach to take an `ng-template` with `matExpansionPanelContent` instead. Can you take another look @jelbourn @josephperrott?

jelbourn commented on src/lib/expansion/expansion-panel.ts:
```diff
@@ -137,6 +152,19 @@ export class MatExpansionPanel extends _MatExpansionPanelMixinBase
+    if (this._lazyContent) {
+      // Render the content as soon as the panel becomes open.
+      this.opened.pipe(
+        startWith(null!),
```
Why does this start with never?

crisbeto replied:
Because the `opened` stream won't emit if the panel is open by default. This ensures that we render any panels that are open on load.

jelbourn commented on src/lib/expansion/expansion.md:
```diff
@@ -99,9 +99,26 @@ panel can be expanded at a given time:
+### Lazy rendering
+
+If you want some content not to be rendered until its expansion panel is opened, you can wrap it in
+an `ng-template` using the `matExpansionPanelContent` directive.
```
How about
`````markdown
By default, the expansion panel content will be initialized even when the panel is closed.
To instead defer initialization until the panel is open, the content should be provided as
an `ng-template`:
```html
<mat-expansion-panel>
  <mat-expansion-panel-header>
    This is the expansion title
  </mat-expansion-panel-header>

  <ng-template matExpansionPanelContent>
    Some deferred content
  </ng-template>
</mat-expansion-panel>
```
`````

crisbeto commented:
Addressed the feedback @jelbourn.

rafaelss95 commented on src/lib/expansion/expansion-panel-content.ts:
```diff
@@ -0,0 +1,20 @@
+  selector: 'ng-template[matExpansionPanelContent]'
+})
+export class MatExpansionPanelContent {
+  constructor(public _template: TemplateRef<any>) {}
```
Just a quick question: Don't you use `_` for `private` variables only?

willshowell replied:
They also use it for "internal use only" properties
